network/proxy/mucp/mucp.go

```diff
@@ -79,155 +84,126 @@ func readLoop(r server.Request, s client.Stream) error {
-			return nil, err
-		}
-
-		for _, service := range services {
-			for _, node := range service.Nodes {
-				addrs = append(addrs, node.Address)
-			}
+	if len(results) > 0 {
+		// update the proxy cache
+		p.Lock()
+		for _, route := range results {
+			p.Routes[service][route.Hash()] = route
 		}
-	}
-
-	// no router addresses available
-	if len(addrs) == 0 {
-		return nil, selector.ErrNoneAvailable
-	}
-
-	var pbRoutes *pb.LookupResponse
-	var gerr error
+		routes = p.Routes[service]
+		p.Unlock()
 
-	// set default client
-	if p.RouterService == nil {
-		p.RouterService = pb.NewRouterService(name, p.Client)
+		return toNodes(routes), nil
 	}
 
-	// TODO: implement backoff and retries
-	for _, addr := range addrs {
-		// call the router
-		proutes, err := p.RouterService.Lookup(context.Background(), &pb.LookupRequest{
-			Query: &pb.Query{
-				Service: service,
-			},
-		}, client.WithAddress(addr))
-		if err != nil {
-			gerr = err
-			continue
-		}
-		// set routes
-		pbRoutes = proutes
-		break
-	}
+	// call the router
+	resp, err := p.RouterService.Lookup(context.Background(), &pb.LookupRequest{
+		Query: &pb.Query{
+			Service: service,
+		},
+	}, p.CallOption)
 
 	// errored out
-	if gerr != nil {
-		return nil, gerr
+	if err != nil {
+		return nil, err
 	}
 
 	// no routes
-	if pbRoutes == nil {
+	if resp == nil || len(resp.Routes) == 0 {
 		return nil, selector.ErrNoneAvailable
 	}
 
 	// convert from pb to []*router.Route
-	for _, r := range pbRoutes.Routes {
-		routes = append(routes, table.Route{
+	for _, r := range resp.Routes {
+		route := table.Route{
 			Service: r.Service,
 			Address: r.Address,
 			Gateway: r.Gateway,
 			Network: r.Network,
 			Link:    r.Link,
 			Metric:  int(r.Metric),
-		})
+		}
+		routes[route.Hash()] = route
 	}
 
 	return toNodes(routes), nil
 }
 
-// ServeRequest honours the server.Router interface
-func (p *Proxy) ServeRequest(ctx context.Context, req server.Request, rsp server.Response) error {
-	// set default client
-	if p.Client == nil {
-		p.Client = client.DefaultClient
+// watchRoutes watches remote router service routes and updates proxy cache
+func (p *Proxy) watchRoutes() {
+	// this is safe to do as the only way watchRoutes returns is
+	// when some error is written into error channel - we want to bail then
+	defer close(p.errChan)
+
+	stream, err := p.RouterService.Watch(context.Background(), &pb.WatchRequest{})
+	if err != nil {
+		p.errChan <- err
+		return
 	}
+	defer stream.Close()
 
+	for {
+		event, err := stream.Recv()
+		if err != nil {
+			p.errChan <- err
+			return
+		}
+
+		p.Lock()
+		route := table.Route{
+			Service: event.Route.Service,
+			Address: event.Route.Address,
+			Gateway: event.Route.Gateway,
+			Network: event.Route.Network,
+			Link:    event.Route.Link,
+			Metric:  int(event.Route.Metric),
+		}
+		if _, ok := p.Routes[route.Service]; !ok {
+			p.Routes[route.Service] = make(map[uint64]table.Route)
+		}
+		p.Routes[route.Service][route.Hash()] = route
```

Comment:
I'm assuming the table event could be a delete or update event along with create so should it act on those?

Reply:
Another great catch; funnily enough I was in the middle of fixing this after making more tests this morning. Yeah, this is by all means a bug and needs to be fixed (the bug fix is in progress rn).